### PR TITLE
Add Azure source e2e tests

### PR DESCRIPTION
**Pull request by cab105**
Migrate the Azure source end to end tests to the triggermesh/triggermesh repo.  These tests require an azure service principal to be created with the following roles assigned at the subscription level:
* `Azure Event Hubs Data Owner`
* `Azure Service Bus Data owner`
* `EventGrid Contributor`
* `Storage Blob Data Owner`

Additionally, when running the ginkgo tests, the following environment variables must be defined:
* `AZURE_SUBSCRIPTION_ID`
* `AZURE_TENANT_ID`
* `AZURE_CLIENT_ID`
* `AZURE_CLIENT_SECRET`
* `AZURE_REGION` is optional and defaults to `uswest2` when omitted

Lastly, this PR updates the makefile to add the tests directory to linting

### CODE REVIEW

**Comment by antoineco** [comment]
I would wait a little bit for the migration. Without an account to run them against, merging would break `main`.

**Comment by cab105** [comment]
> I would wait a little bit for the migration. Without an account to run them against, merging would break `main`.

That's a good point. There is also at least one more test on the way as well so I'll keep it open for now.

File: go.mod
Comment (antoineco) [inline go.mod:59]: It looks like we missed `/v2` some in some import and Go automatically assumed we want to bring back Ginkgo v1.

**Comment by sebgoa** [comment]
Heather and I are trying to get an org account on Azure, it is a bit of a maze unfortunately.

File: go.mod
Comment (cab105) [reply go.mod]: Yeah this was me fighting my IDE that insisted on removing the `_ github.com/onsi/ginkgo/v2` import I had originally. Should be resolved now.

**Comment by cab105** [comment]
Re-opening this PR.  The third-party license stems from three sub directories within the `github.com/Azure/azure-sdk-for-go` repo and should still be under the MIT license. Please advise on how best to continue.

**Review by antoineco — COMMENTED** [review]
Looking good!

I would suggest keeping this a draft to avoid accidental merges until our Azure account is ready and configured with a Service Principal for running tests.

**Comment by cab105** [comment]
> Looking good!
> 
> I would suggest keeping this a draft to avoid accidental merges until our Azure account is ready and configured with a Service Principal for running tests.

It already is. Made sure to take care of it earlier today, and verified it prior to pushing out the latest batch of changes.

File: Makefile
Comment (cab105) [reply Makefile:151]: Good question on `$GOPKGS`. Think this was to exclude some of the stub code used to generate new targets when things were closed. Can remove it here though.

File: go.mod
Comment (antoineco) [reply go.mod:26]: You're updating from `v61.0.0` to `v61.1.0`, which is unlikely to have an influence on the other modules. Besides these other modules are versioned independently from the "main" SDK module (they just happen to be pushed to the same repo).

Comment (antoineco) [reply go.mod:26]: Since we're talking about it: #429 😅  (I didn't trigger it)

**Comment by cab105** [comment]
> @cab105 regarding the vendored OSS licenses, here is the procedure, in two commands, to update them:
> 
> https://github.com/triggermesh/triggermesh/blob/67e8d5c90b172de8187ca9ee46f6ef0bb7f30f3a/.github/workflows/licenses.yaml#L77
> 
> https://github.com/triggermesh/triggermesh/blob/67e8d5c90b172de8187ca9ee46f6ef0bb7f30f3a/.github/workflows/licenses.yaml#L81
> 
> **edit:** my bad, I see the errors now:
> 
> ```
> F0112 17:50:58.317894   24533 main.go:43] one or more libraries have an incompatible/unknown license: map["unknown":["github.com/Azure/azure-sdk-for-go/sdk/internal/log" "github.com/Azure/azure-sdk-for-go/sdk/internal/errorinfo" "github.com/Azure/azure-sdk-for-go/sdk/internal/diag" "github.com/Azure/azure-sdk-for-go/sdk/internal/uuid"]]
> ```

My understanding with that check is by using the `--force` flag it will delete the `LICENSES/vendor` directory and rebuild it based on the contents of `go.mod` which means the check will still fail. For the four paths you cite as they don't have a `LICENSE.txt` file since it exists in `github.com/Azure/azure-sdk-for-go`.

File: go.mod
Comment (cab105) [reply go.mod:26]: Okay now that is funny. Will wait for dependabot to merge it first and rebase.

**Comment by antoineco** [comment]
@cab105 I know how to fix it. The issue was known and a license file was included in `sdk/internal@v0.8.3`: https://github.com/Azure/azure-sdk-for-go/tree/sdk/internal/v0.8.3/sdk/internal

All we have to do is ensure that we use dependencies that import that version instead of `v0.8.1`.

**Comment by cab105** [comment]
> @cab105 I know how to fix it. The issue was known and a license file was included in `sdk/internal@v0.8.3`: https://github.com/Azure/azure-sdk-for-go/tree/sdk/internal/v0.8.3/sdk/internal
> 
> All we have to do is import that version instead of `v0.8.1`:
> 
> ```
> $ go get github.com/Azure/azure-sdk-for-go/sdk/internal@v0.8.3
> go: downloading github.com/Azure/azure-sdk-for-go/sdk/internal v0.8.3
> go get: upgraded github.com/Azure/azure-sdk-for-go/sdk/internal v0.8.1 => v0.8.3
> ```
> 
> ```
> $ go-licenses save ./... --save_path LICENSES/vendor/ --force
> 👍
> ```

Okay that's a slightly cleaner approach than the other workaround I found which is to have the license check call `go mod vendor` to pull the dependencies into a vendor directory and then run the license check, which will pass.

**Review by antoineco — COMMENTED** [review]
LGTM but remember: we need to add secrets to GitHub - and ideally ensure none of the tests are flaky by running then ~10 times - before we can merge.

**Comment by cab105** [comment]
@antoineco Wouldn't the secrets need to go into CircleCI since that is where the testing happens?

File: test/e2e/framework/azure/eventhub.go
Comment (antoineco) [inline test/e2e/framework/azure/eventhub.go:91]: I was running the tests locally with the minimum set of permissions possible, and accidentally stumbled upon an error here because I hadn't assigned `Microsoft.EventHub/namespaces/authorizationRules/listKeys/action`.

Long story short: I was wondering why we do all this trickery to read a "root" key and assemble a connection string, when it is absolutely possible to use the exact same credentials as above to obtain an Event Hubs client? (we do it [in the reconciler](https://github.com/triggermesh/triggermesh/blob/33b9cc4196de99621a93966cf14c4eabffee6e3c/pkg/sources/client/azure/eventgrid/client.go#L73-L74), as an example)

Comment (cab105) [reply test/e2e/framework/azure/eventhub.go:91]: Think this was a holdover from an earlier version (before realizing I needed to pass `/v3` for the eventhub sdk) which required the key in addition to the connection string. Will clean this up.

**Review by antoineco — APPROVED** [review]
The repo is set up with the required secrets.

It would be nice to clean up the strange thing with the Root key, but regardless, everything seems to work as expected.

**Pull request merged**
